kernel/src/hil/time.rs

```diff
@@ -1,33 +1,270 @@
+/// value and returned to 0.
+pub trait OverflowClient {
+    fn overflow(&self);
 }
 
-pub trait Counter<W = u32>: Time<W> {
+/// Represents a free-running hardware counter that can be started and stopped.
+pub trait Counter<'a>: Time {
+    /// Specify the callback for when the counter overflows its maximum
+    /// value (defined by `Ticks`). If there was a previously registered
+    /// callback this call replaces it.
+    fn set_overflow_client(&'a self, client: &'a dyn OverflowClient);
+
+    /// Starts the free-running hardware counter. Valid `ReturnCode` values are:
+    ///   - `ReturnCode::SUCCESS`: the counter is now running
+    ///   - `ReturnCode::EOFF`: underlying clocks or other hardware resources
+    ///   are not on, such that the counter cannot start.
+    ///   - `ReturnCode::FAIL`: unidentified failure, counter is not running.
+    /// After a successful call to `start`, `is_running` MUST return true.    
     fn start(&self) -> ReturnCode;
+
+    /// Stops the free-running hardware counter. Valid `ReturnCode` values are:
+    ///   - `ReturnCode::SUCCESS`: the counter is now stopped.
+    ///   - `ReturnCode::EBUSY`: the counter is in use in a way that means it
+    ///   cannot be stopped and is busy.
+    ///   - `ReturnCode::FAIL`: unidentified failure, counter is running.
+    /// After a successful call to `stop`, `is_running` MUST return false.        
     fn stop(&self) -> ReturnCode;
+
+    /// Resets the counter to 0. If the counter is currently running, this
+    /// may require stopping and restarting it. Valid `ReturnCode` values are:
+    ///    - `ReturnCode::SUCCESS`: the counter was reset to 0.
+    ///    - `ReturnCode::FAIL`: the counter was not reset to 0.    
+    fn reset(&self) -> ReturnCode;
+
+    /// Returns whether the counter is currently running.
     fn is_running(&self) -> bool;
 }
 
-/// Trait to represent clock frequency in Hz
-///
-/// This trait is used as an associated type for `Alarm` so clients can portably
-/// convert native cycles to real-time values.
-pub trait Frequency {
-    /// Returns frequency in Hz.
-    fn frequency() -> u32;
+/// Callback handler for when an Alarm fires (a `Counter` reaches a specific
+/// value).
+pub trait AlarmClient {
+    /// Callback indicating the alarm time has been reached. The alarm
+    /// MUST be disabled when this is called. If a new alarm is needed,
+    /// the client can call `Alarm::set_alarm`.
+    fn alarm(&self);
+}
+
+/// Interface for receiving notification when a particular time
+/// (`Counter` value) is reached. Clients use the
+/// [`AlarmClient`](trait.AlarmClient.html) trait to signal when the
+/// counter has reached a pre-specified value set in
+/// [`set_alarm`](#tymethod.set_alarm). Alarms are intended for
+/// low-level time needs that require precision (i.e., firing on a
+/// precise clock tick). Software that needs more functionality
+/// but can tolerate some jitter should use the `Timer` trait
+/// instead.
```

Comment:
My observation is that currently `Timer` is barely used throughout Tock, despite there being many instances where jitter can be tolerated. What is the benefit of using `Timer` instead of an alarm? Is it just a cleaner interface, or are there performance benefits to using `Timer` when precision is not needed?

Reply:
There are lots of cases where we use Alarm where Timer is a better bet; typical cases are periodic operations (such as sensing, sending a packet). It happens to be historically that we didn't have a solid Timer implementation so everyone used Alarm.